### ISSUE TITLE
How to use kv-cache in my own onnx to convert engine?

### ISSUE THREAD

**Issue by triple-Mu** [NVIDIA/TensorRT#3036]
I want to speed a nlp model. 
I notice trt8.6 has supported kv-cache.
By the demo bart seems need 2 models non-kv and kv model to make a full pipeline.
But it will have so much inputs maybe 48(24 attns) and so much outputs.
It also seems that some op fusion will fail when there are so much inputs and outputs.
Could you help me to support kv-cache in one model or give me an example.
Thanks.

**Comment by zerollzeng** [comment]
@ttyio ^ ^

**Comment by ttyio** [comment]
@triple-Mu , the extra inputs is for the saving and loading of KV cache, this is per encoder, so we need these extra input/output. And we will have more examples in future release, thanks!

**Comment by triple-Mu** [comment]
> @triple-Mu , the extra inputs is for the saving and loading of KV cache, this is per encoder, so we need these extra input/output. And we will have more examples in future release, thanks!

But now I need to prepare two onnx for converting.
One is non-kv, 4 inputs 49 outputs(1+48 kv)
One is kv, 52(4+48kv) inputs and 49 outputs(1+48 kv).
How could I using dynamic number of inputs?

**Comment by ttyio** [comment]
@triple-Mu , what about create 2 context, one for context and the other for decoder, and use different number of inputs for them.

**Comment by triple-Mu** [comment]
> @triple-Mu , what about create 2 context, one for context and the other for decoder, and use different number of inputs for them.

But I can not build an engine with two different number of inputs.
Could you please give me an example?

**Comment by ttyio** [comment]
We need 2 engines given they are 2 models.

**Comment by nvluxiaoz** [comment]
TensorRT does not support dynamic number of inputs, while torch.onnx.export uses static operation traces, which adds extra difficulties for supporting kv cache for NLP models. The example of BART has to account the factor for the 1st decoding session does not have KV Cache input, so we need to generate the first iteration of self_attention kv cache and cross_attention kv cache, so 2 engines are required. 

There are definitely optimizations that TensorRT (or users) can do, is to use TensorRT's zero-tensor feature, which is actually showcased in [T5](https://github.com/NVIDIA/TensorRT/tree/release/8.6/demo/HuggingFace/T5) example. The logic behind:
(1) encoder: input_ids -> encoder_hidden_states
(2) decoder: input_ids + encoder_hidden_states + past_key_values.decoder(self_attention) * 24 range from 0-max_length + past_key_values.encoder(cross_attention) * 24 -> logits + present_key_values.decoder * 24
(3) cross_attention_generator: example here https://github.com/NVIDIA/TensorRT/blob/release/8.6/demo/HuggingFace/T5/export.py#L194. encoder_hidden_states -> past_key_values.encoder(cross_attention) * 24, which does not change, and we cache the cross_attention_kv_cache into the binding.

The logical difficulty of implementing dynamic number of inputs is:
```
if not cross_attention_cache:
     compute cross_attention_cache

use_cross_attention_cache
```

**Comment by ttyio** [comment]
closing since there is detail answer, thanks all!
